### PR TITLE
transaction: always lock keys with presume not exists flag

### PR DESCRIPTION
**Pull request by coocood**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/19063

### What is changed and how it works?

always lock for key with presume not exists flag.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

- always lock keys with presume not exists flag

### CODE REVIEW

**Comment by coocood** [comment]
@ekexium PTAL

File: session/txn.go [hunk 1]
```diff
@@ -294,20 +294,23 @@ func (st *TxnState) KeysNeedToLock() ([]kv.Key, error) {
...
 }
 
-func keyNeedToLock(k, v []byte) bool {
+func keyNeedToLock(k, v []byte, flags kv.KeyFlags) bool {
 	isTableKey := bytes.HasPrefix(k, tablecodec.TablePrefix())
 	if !isTableKey {
 		// meta key always need to lock.
 		return true
 	}
+	if flags.HasPresumeKeyNotExists() {
```
Comment (cfzjywxk) [inline session/txn.go:311]: Should we use `flags.HasNeedCheckExists()` as it's statement level flags.

Comment (coocood) [reply session/txn.go:311]: If the flag has `PresumeKeyNotExists` but doesn't has `NeedCheckExists` flag, the key must have been locked already, locking it again doesn't matter.

**Comment by cfzjywxk** [comment]
LGTM

**Pull request merged**